Source/Core/VideoCommon/BPMemory.h

```diff
@@ -104,140 +111,209 @@ enum
 // Tev/combiner things
 
 // TEV scaling type
-enum : u32
+enum class TevScale : u32
 {
-  TEVSCALE_1 = 0,
-  TEVSCALE_2 = 1,
-  TEVSCALE_4 = 2,
-  TEVDIVIDE_2 = 3
+  Scale1 = 0,
+  Scale2 = 1,
+  Scale4 = 2,
+  Divide2 = 3
 };
-
-enum : u32
+template <>
+struct fmt::formatter<TevScale> : EnumFormatter<TevScale::Divide2>
 {
-  TEVCMP_R8 = 0,
-  TEVCMP_GR16 = 1,
-  TEVCMP_BGR24 = 2,
-  TEVCMP_RGB8 = 3
+  formatter() : EnumFormatter({"1", "2", "4", "0.5"}) {}
 };
 
 // TEV combiner operator
-enum : u32
-{
-  TEVOP_ADD = 0,
-  TEVOP_SUB = 1,
-  TEVCMP_R8_GT = 8,
-  TEVCMP_R8_EQ = 9,
-  TEVCMP_GR16_GT = 10,
-  TEVCMP_GR16_EQ = 11,
-  TEVCMP_BGR24_GT = 12,
-  TEVCMP_BGR24_EQ = 13,
-  TEVCMP_RGB8_GT = 14,
-  TEVCMP_RGB8_EQ = 15,
-  TEVCMP_A8_GT = TEVCMP_RGB8_GT,
-  TEVCMP_A8_EQ = TEVCMP_RGB8_EQ
+enum class TevOp : u32
+{
+  Add = 0,
+  Sub = 1,
+};
+template <>
+struct fmt::formatter<TevOp> : EnumFormatter<TevOp::Sub>
+{
+  formatter() : EnumFormatter({"Add", "Subtract"}) {}
+};
+
+enum class TevCompareMode : u32
+{
+  R8 = 0,
+  GR16 = 1,
+  BGR24 = 2,
+  RGB8 = 3,
+  A8 = RGB8,
+};
+template <>
+struct fmt::formatter<TevCompareMode> : EnumFormatter<TevCompareMode::RGB8>
+{
+  formatter() : EnumFormatter({"R8", "GR16", "BGR24", "RGB8 / A8"}) {}
+};
+
+enum class TevComparison : u32
+{
+  GT = 0,
+  EQ = 1,
+};
+template <>
+struct fmt::formatter<TevComparison> : EnumFormatter<TevComparison::EQ>
+{
+  formatter() : EnumFormatter({"Greater than", "Equal to"}) {}
 };
 
 // TEV color combiner input
-enum : u32
-{
-  TEVCOLORARG_CPREV = 0,
-  TEVCOLORARG_APREV = 1,
-  TEVCOLORARG_C0 = 2,
-  TEVCOLORARG_A0 = 3,
-  TEVCOLORARG_C1 = 4,
-  TEVCOLORARG_A1 = 5,
-  TEVCOLORARG_C2 = 6,
-  TEVCOLORARG_A2 = 7,
-  TEVCOLORARG_TEXC = 8,
-  TEVCOLORARG_TEXA = 9,
-  TEVCOLORARG_RASC = 10,
-  TEVCOLORARG_RASA = 11,
-  TEVCOLORARG_ONE = 12,
-  TEVCOLORARG_HALF = 13,
-  TEVCOLORARG_KONST = 14,
-  TEVCOLORARG_ZERO = 15
+enum class TevColorArg : u32
+{
+  PrevColor = 0,
+  PrevAlpha = 1,
+  Color0 = 2,
+  Alpha0 = 3,
+  Color1 = 4,
+  Alpha1 = 5,
+  Color2 = 6,
+  Alpha2 = 7,
+  TexColor = 8,
+  TexAlpha = 9,
+  RasColor = 10,
+  RasAlpha = 11,
+  One = 12,
+  Half = 13,
+  Konst = 14,
+  Zero = 15
+};
+template <>
+struct fmt::formatter<TevColorArg> : EnumFormatter<TevColorArg::Zero>
+{
+  static constexpr array_type names = {
+      "prev.rgb", "prev.aaa", "c0.rgb",  "c0.aaa",  "c1.rgb", "c1.aaa", "c2.rgb",    "c2.aaa",
```

Comment:
Is there a reason you went with `aaa`?  Could just be my time with shader languages but I was expecting `prev.a`

Reply:
These names [were already present before](https://github.com/dolphin-emu/dolphin/blob/9d0983c9c988f81c73c9a4ed61835ca9ac20db06/Source/Core/VideoCommon/BPStructs.cpp#L1289-L1292), so I didn't think about them too much.

I'm not super familiar with shader languages; is something like `color.rgb = prev.a` legal (meaning the same thing as `color.rgb = prev.aaa`)?  I know you can write `color.rgb = float3(prev.a)` (or something like that) at least.

Reply:
Ah, looking at the previous code, I think the triple-a is probably intentional to show that the value is a 3 float component all filled with the alpha.  Sorry, wasn't thinking about this very clearly last night.  This can be marked resolved.